File: apps/api/src/digital-id-cards/digital-id-cards.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { randomBytes } from 'node:crypto';
import { SupabaseService } from '../supabase/supabase.service';
import { StorageService } from '../storage/storage.service';
import { SijilliErrors } from '../common/errors/error-envelope';
import { SijilliRequestUser } from '../auth/types';
import { DigitalIdNumberService } from '../citizens/digital-id-number.service';
import { NfcKeyStoreService } from '../nfc/nfc-key-store.service';
import { sha256OfStorageObject } from '../common/utils/sha256';
import { SsiService } from '../ssi/ssi.service';
import { IssueCardDto } from './dto/issue-card.dto';
import { FreezeCardDto, ReissueCardDto, RevokeCardDto } from './dto/card-action.dto';
import { ListDigitalIdsQuery } from './dto/list-cards.dto';

const CARD_SELECT = `
  id, citizen_id, digital_id_number, card_serial,
  nfc_uid, nfc_signature_key_id,
  did, did_doc, wallet_endpoint,
  issued_at, issued_by_officer_id, expires_at,
  status, revoked_at, revoked_reason,
  photo_hash, data_hash,
  last_nfc_counter, last_nfc_tap_at,
  created_at, updated_at
`;

const PHOTO_SHA256_RE = /^[0-9a-fA-F]{64}$/;

export interface IssueCardResult {
  card: Record<string, unknown>;
  // The plaintext NFC keys are returned exactly once — the issuer station
  // writes them to the chip and never stores them. After this response,
  // the keys live only inside nfc_card_secrets, encrypted with KMS.
  nfc_keys: {
    meta_read_key_hex: string;
    sdm_file_read_key_hex: string;
    kms_key_id: string;
  };
  // The URL template that should be encoded into NDEF record 2 of the
  // chip. The chip fills in the {p} and {c} placeholders on every tap.
  sun_url_template: string;
}

@Injectable()
export class DigitalIdCardsService {
  private readonly logger = new Logger(DigitalIdCardsService.name);

  constructor(
    private readonly supabase: SupabaseService,
    private readonly storage: StorageService,
    private readonly numbers: DigitalIdNumberService,
    private readonly keyStore: NfcKeyStoreService,
    private readonly ssi: SsiService,
  ) {}

  // ----------------------------------------------------------------------
  // LIST — cursor-paginated, optionally filtered by status / id prefix.
  // The joined citizen lets the admin UI render full Arabic names without
  // an extra round-trip per row.
  // ----------------------------------------------------------------------
  async list(q: ListDigitalIdsQuery, actor: SijilliRequestUser) {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    let query = this.supabase.admin
      .from('digital_id_cards')
      .select(
        `${CARD_SELECT}, citizen:citizens(id, first_name_ar, father_name_ar, family_name_ar, region_id, phone)`,
      )
      .order('issued_at', { ascending: false })
      .order('id', { ascending: false })
      .limit(q.limit + 1);

    if (q.status) query = query.eq('status', q.status);
    if (q.cursor) query = query.lt('issued_at', q.cursor);
    if (q.q && q.q.trim().length >= 2) {
      const safe = q.q.replace(/[%_]/g, '\\$&').trim();
      query = query.ilike('digital_id_number', `%${safe}%`);
    }

    const { data, error } = await query;
    if (error) {
      throw SijilliErrors.upstream(`Failed to list digital ID cards: ${error.message}`);
    }

    const items = (data ?? []) as Array<Record<string, unknown>>;
    let next_cursor: string | null = null;
    if (items.length > q.limit) {
      const overflow = items.pop()!;
      next_cursor = (overflow.issued_at as string) ?? null;
    }
    return { items, next_cursor };
  }

  // ----------------------------------------------------------------------
  // ISSUE — full new-card flow.
  // ----------------------------------------------------------------------
  async issue(dto: IssueCardDto, actor: SijilliRequestUser): Promise<IssueCardResult> {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    // 1. Confirm the citizen exists and is active.
    const citizen = await this.supabase.admin
      .from('citizens')
      .select('id, region_id, is_active, photo_path')
      .eq('id', dto.citizen_id)
      .maybeSingle();
    if (citizen.error) throw SijilliErrors.upstream(citizen.error.message);
    if (!citizen.data || !citizen.data.is_active) throw SijilliErrors.notFound('المواطن');

    // 2. Refuse if the citizen already has an active card — caller should
    //    use /reissue instead.
    const existing = await this.supabase.admin
      .from('digital_id_cards')
      .select('id')
      .eq('citizen_id', dto.citizen_id)
      .eq('status', 'active')
      .maybeSingle();
    if (existing.data) {
      throw SijilliErrors.conflict(
        'يوجد بطاقة فعّالة لهذا المواطن. استخدم إعادة الإصدار بدلاً من إصدار جديد.',
        'Citizen already has an active card; use /reissue.',
      );
    }

    // 3. Resolve the photo hash.
    const photoHash = await this.resolvePhotoHash(dto, citizen.data.photo_path ?? null);

    // 4. Allocate the digital ID number (real Luhn check digit).
    const digitalIdNumber = await this.numbers.next(
      dto.region_code,
      dto.year ?? new Date().getFullYear(),
    );

    // 5. Insert the card row in 'active' state.
    const expiresAt = addYears(new Date(), dto.validity_years ?? 5);
    const cardSerial = `LY-${randomHexUpper(12)}`;

    const insert = await this.supabase.admin
      .from('digital_id_cards')
      .insert({
        citizen_id: dto.citizen_id,
        digital_id_number: digitalIdNumber,
        card_serial: cardSerial,
        photo_hash: photoHash,
        issued_by_officer_id: actor.officer_id,
        expires_at: expiresAt.toISOString(),
        status: 'active',
      })
      .select(CARD_SELECT)
      .single();

    if (insert.error || !insert.data) {
      if (insert.error?.code === '23505') {
        throw SijilliErrors.conflict(
          'تعارض في رقم البطاقة أو الرقم الرقمي.',
          'Conflict on card_serial or digital_id_number.',
        );
      }
      throw SijilliErrors.upstream(`Failed to insert card: ${insert.error?.message}`);
    }

    // 6. Mint and persist the NFC keys (returned once to the caller).
    const keys = await this.keyStore.mintForCard(insert.data.id as string);

    // 7. History entry.
    await this.supabase.admin.from('id_issuance_history').insert({
      citizen_id: dto.citizen_id,
      card_id: insert.data.id,
      action: 'issued',
      reason: null,
      officer_id: actor.officer_id,
    });

    // 8. Issue the DigitalId VC. Failure here doesn't block card issuance —
    // the workflow can retry later — but we log it for ops visibility.
    const card = await this.attachVc(insert.data.id as string, insert.data);

    return {
      card,
      nfc_keys: {
        meta_read_key_hex: keys.metaReadKey.toString('hex').toUpperCase(),
        sdm_file_read_key_hex: keys.sdmFileReadKey.toString('hex').toUpperCase(),
        kms_key_id: 'local:v1',
      },
      sun_url_template:
        (process.env.NFC_SUN_BASE_URL ?? 'https://verify.sijilli.ly/v') + '?p={picc}&c={cmac}',
    };
  }

  // ----------------------------------------------------------------------
  // FREEZE — reversible. Card stops verifying tap responses but keys are kept.
  // ----------------------------------------------------------------------
  async freeze(cardId: string, dto: FreezeCardDto, actor: SijilliRequestUser) {
    return this.transition(cardId, 'frozen', dto.reason, actor, 'frozen');
  }

  // ----------------------------------------------------------------------
  // REVOKE — terminal. Card secrets are kept (for audit) but status is final.
  // ----------------------------------------------------------------------
  async revoke(cardId: string, dto: RevokeCardDto, actor: SijilliRequestUser) {
    return this.transition(cardId, 'revoked', dto.reason, actor, 'revoked');
  }

  // ----------------------------------------------------------------------
  // REISSUE — revoke the existing card and issue a new one for the same
  // citizen. By default keeps the same digital_id_number (number is the
  // identity, the card is the credential).
  // ----------------------------------------------------------------------
  async reissue(cardId: string, dto: ReissueCardDto, actor: SijilliRequestUser): Promise<IssueCardResult> {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    const old = await this.supabase.admin
      .from('digital_id_cards')
      .select('id, citizen_id, digital_id_number, photo_hash, expires_at, status')
      .eq('id', cardId)
      .maybeSingle();
    if (old.error) throw SijilliErrors.upstream(old.error.message);
    if (!old.data) throw SijilliErrors.notFound('البطاقة');

    // Revoke the old card.
    await this.transition(cardId, 'revoked', `إعادة إصدار: ${dto.reason}`, actor, 'revoked');

    // Issue a new one. Reuse the digital_id_number unless asked not to.
    const region = parseRegionFromDigitalId(old.data.digital_id_number as string);
    const year = new Date().getFullYear();
    const digitalIdNumber =
      dto.keep_digital_id_number === false
        ? await this.numbers.next(region, year)
        : (old.data.digital_id_number as string);

    const cardSerial = `LY-${randomHexUpper(12)}`;
    const expiresAt = addYears(new Date(), 5);

    const insert = await this.supabase.admin
      .from('digital_id_cards')
      .insert({
        citizen_id: old.data.citizen_id,
        digital_id_number: digitalIdNumber,
        card_serial: cardSerial,
        photo_hash: old.data.photo_hash,
        issued_by_officer_id: actor.officer_id,
        expires_at: expiresAt.toISOString(),
        status: 'active',
      })
      .select(CARD_SELECT)
      .single();

    if (insert.error || !insert.data) {
      throw SijilliErrors.upstream(`Failed to insert reissued card: ${insert.error?.message}`);
    }

    const keys = await this.keyStore.mintForCard(insert.data.id as string);

    await this.supabase.admin.from('id_issuance_history').insert({
      citizen_id: old.data.citizen_id,
      card_id: insert.data.id,
      action: 're-issued',
      reason: dto.reason,
      officer_id: actor.officer_id,
    });

    const card = await this.attachVc(insert.data.id as string, insert.data);

    return {
      card,
      nfc_keys: {
        meta_read_key_hex: keys.metaReadKey.toString('hex').toUpperCase(),
        sdm_file_read_key_hex: keys.sdmFileReadKey.toString('hex').toUpperCase(),
        kms_key_id: 'local:v1',
      },
      sun_url_template:
        (process.env.NFC_SUN_BASE_URL ?? 'https://verify.sijilli.ly/v') + '?p={picc}&c={cmac}',
    };
  }

  // Issue the DigitalId VC for a freshly-active card and stamp the
  // wallet's DID onto the card row. Errors are logged, not thrown — VC
  // issuance is best-effort and recoverable via /digital-id-cards/:id/vc/retry
  // (Phase 12). Returns the (possibly updated) card row.
  private async attachVc(
    cardId: string,
    initialCard: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    try {
      const vc = await this.ssi.issueDigitalIdVc({ card_id: cardId });
      const updated = await this.supabase.admin
        .from('digital_id_cards')
        .update({ did: vc.did })
        .eq('id', cardId)
        .select(CARD_SELECT)
        .single();
      if (updated.error || !updated.data) {
        this.logger.warn(`Failed to stamp DID on card ${cardId}: ${updated.error?.message}`);
        return initialCard;
      }
      return updated.data;
    } catch (err) {
      this.logger.error(`VC issuance failed for card ${cardId}: ${(err as Error).message}`);
      return initialCard;
    }
  }

  // ----------------------------------------------------------------------
  // Helpers
  // ----------------------------------------------------------------------

  private async transition(
    cardId: string,
    nextStatus: 'frozen' | 'revoked',
    reason: string,
    actor: SijilliRequestUser,
    historyAction: string,
  ) {
    if (!actor.officer_id) throw SijilliErrors.forbidden();

    const before = await this.supabase.admin
      .from('digital_id_cards')
      .select('id, citizen_id, status')
      .eq('id', cardId)
      .maybeSingle();
    if (before.error) throw SijilliErrors.upstream(before.error.message);
    if (!before.data) throw SijilliErrors.notFound('البطاقة');
    if (before.data.status === 'revoked') {
      throw SijilliErrors.conflict(
        'البطاقة مُلغاة بالفعل ولا يمكن تعديل حالتها.',
        'Card is already revoked.',
      );
    }
    if (nextStatus === 'frozen' && before.data.status === 'frozen') {
      throw SijilliErrors.conflict('البطاقة مجمّدة بالفعل.', 'Card is already frozen.');
    }

    const patch: Record<string, unknown> = { status: nextStatus };
    if (nextStatus === 'revoked') {
      patch.revoked_at = new Date().toISOString();
      patch.revoked_reason = reason;
    }

    const updated = await this.supabase.admin
      .from('digital_id_cards')
      .update(patch)
      .eq('id', cardId)
      .select(CARD_SELECT)
      .single();
    if (updated.error || !updated.data) {
      throw SijilliErrors.upstream(`Failed to update card: ${updated.error?.message}`);
    }

    await this.supabase.admin.from('id_issuance_history').insert({
      citizen_id: before.data.citizen_id,
      card_id: cardId,
      action: historyAction,
      reason,
      officer_id: actor.officer_id,
    });

    return updated.data;
  }

  private async resolvePhotoHash(dto: IssueCardDto, citizenPhotoPath: string | null): Promise<string> {
    if (dto.photo_sha256) {
      if (!PHOTO_SHA256_RE.test(dto.photo_sha256)) {
        throw SijilliErrors.validation(
          'بصمة الصورة غير صالحة (يجب أن تكون 64 حرفاً سادس عشر).',
          'photo_sha256 must be 64 hex characters.',
        );
      }
      return dto.photo_sha256.toLowerCase();
    }

    const path = dto.photo_path ?? citizenPhotoPath;
    if (!path) {
      throw SijilliErrors.validation(
        'يجب توفير صورة المواطن أو بصمتها قبل إصدار البطاقة.',
        'Either photo_path or photo_sha256 is required.',
      );
    }

    return sha256OfStorageObject(this.storage, dto.photo_bucket ?? 'citizen_photos', path);
  }
}

// ----------------------------------------------------------------------
// Pure helpers
// ----------------------------------------------------------------------

function addYears(d: Date, years: number): Date {
  const out = new Date(d.getTime());
  out.setFullYear(out.getFullYear() + years);
  return out;
}

function randomHexUpper(bytes: number): string {
  return randomBytes(bytes).toString('hex').toUpperCase();
}

// LY-RR-YYYY-SSSSSS-C → "RR"
function parseRegionFromDigitalId(id: string): string {
  const m = /^LY-([0-9]{2,4})-/.exec(id);
  if (!m) throw SijilliErrors.upstream(`Cannot parse region from digital ID: ${id}`);
  return m[1];
}

File: apps/api-dotnet/DigitalIdCards/DigitalIdCardsService.cs
using System.Security.Cryptography;
using System.Text.RegularExpressions;
using Microsoft.EntityFrameworkCore;
using Sarh.Api.Auth;
using Sarh.Api.Common;
using Sarh.Api.Common.Errors;
using Sarh.Api.Data;
using Sarh.Api.Data.Entities;
using Sarh.Api.Nfc;
using Sarh.Api.Notifications;

namespace Sarh.Api.DigitalIdCards;

public sealed partial class DigitalIdCardsService(
    SarhDbContext db,
    DigitalIdNumberService numbers,
    NfcKeyStoreService keyStore,
    NotificationsService notifications,
    Sarh.Api.Ssi.ISsiService ssi,
    IConfiguration config,
    ILogger<DigitalIdCardsService> log)
{
    private static readonly Regex PhotoSha256Re = new("^[0-9a-fA-F]{64}$");

    // ---------------- LIST ----------------
    public async Task<CursorPage<CardView>> ListAsync(ListCardsQuery q, CurrentUser actor, CancellationToken ct)
    {
        IQueryable<DigitalIdCard> query = db.DigitalIdCards.AsNoTracking();

        // Citizens can only see their own card. Officers see what they query.
        if (actor.OfficerId is null)
        {
            if (actor.CitizenId is null) throw SarhException.Forbidden();
            query = query.Where(c => c.CitizenId == actor.CitizenId.Value);
        }
        else if (q.CitizenId is Guid cid)
        {
            query = query.Where(c => c.CitizenId == cid);
        }

        if (!string.IsNullOrEmpty(q.Status)) query = query.Where(c => c.Status == q.Status);
        if (!string.IsNullOrWhiteSpace(q.Cursor) && DateTimeOffset.TryParse(q.Cursor, out var cursorTs))
            query = query.Where(c => c.IssuedAt < cursorTs);
        if (!string.IsNullOrWhiteSpace(q.Q) && q.Q.Trim().Length >= 2)
        {
            var pat = "%" + q.Q.Trim().Replace("[", "[[]").Replace("%", "[%]").Replace("_", "[_]") + "%";
            query = query.Where(c => EF.Functions.Like(c.DigitalIdNumber, pat));
        }

        var rows = await query
            .OrderByDescending(c => c.IssuedAt)
            .ThenByDescending(c => c.Id)
            .Take(q.Limit + 1)
            .ToListAsync(ct);

        string? nextCursor = null;
        if (rows.Count > q.Limit)
        {
            nextCursor = rows[q.Limit].IssuedAt.ToString("o");
            rows = rows.Take(q.Limit).ToList();
        }

        // Fetch citizen summaries for the page in a single round-trip.
        var citizenIds = rows.Select(r => r.CitizenId).Distinct().ToList();
        var citizenMap = await db.Citizens.AsNoTracking()
            .Where(c => citizenIds.Contains(c.Id))
            .Select(c => new CardCitizenSummary
            {
                Id = c.Id,
                FirstNameAr = c.FirstNameAr,
                FatherNameAr = c.FatherNameAr,
                FamilyNameAr = c.FamilyNameAr,
                RegionId = c.RegionId,
                Phone = c.Phone,
            })
            .ToDictionaryAsync(c => c.Id, ct);

        return new CursorPage<CardView>
        {
            Items = rows.Select(c => CardView.From(c, citizenMap.GetValueOrDefault(c.CitizenId))).ToList(),
            NextCursor = nextCursor,
        };
    }

    // ---------------- ISSUE ----------------
    public async Task<IssueCardResult> IssueAsync(IssueCardDto dto, CurrentUser actor, CancellationToken ct)
    {
        if (actor.OfficerId is null) throw SarhException.Forbidden();

        var citizen = await db.Citizens.FirstOrDefaultAsync(c => c.Id == dto.CitizenId, ct);
        if (citizen is null || !citizen.IsActive) throw SarhException.NotFound("المواطن", "Citizen");

        var hasActive = await db.DigitalIdCards.AsNoTracking()
            .AnyAsync(c => c.CitizenId == dto.CitizenId && c.Status == "active", ct);
        if (hasActive)
            throw SarhException.Conflict(
                "يوجد بطاقة فعّالة لهذا المواطن. استخدم إعادة الإصدار بدلاً من إصدار جديد.",
                "Citizen already has an active card; use /reissue.");

        // Persist the upload onto the citizen so /citizens/:id/photo can serve it.
        if (!string.IsNullOrEmpty(dto.PhotoBucket) && !string.IsNullOrEmpty(dto.PhotoPath))
            citizen.PhotoPath = $"{dto.PhotoBucket}/{dto.PhotoPath}";

        var photoHash = ResolvePhotoHash(dto, citizen.PhotoPath);

        var year = dto.Year ?? DateTime.UtcNow.Year;
        var digitalIdNumber = await numbers.NextAsync(dto.RegionCode, year, ct);

        var validityYears = dto.ValidityYears ?? 5;
        var expiresAt = DateTimeOffset.UtcNow.AddYears(validityYears);
        var cardSerial = $"LY-{RandomHexUpper(12)}";

        var card = new DigitalIdCard
        {
            Id = Guid.NewGuid(),
            CitizenId = dto.CitizenId,
            DigitalIdNumber = digitalIdNumber,
            CardSerial = cardSerial,
            PhotoHash = photoHash,
            DataHash = IdentityHash.Compute(citizen, digitalIdNumber),
            IssuedByOfficerId = actor.OfficerId,
            IssuedAt = DateTimeOffset.UtcNow,
            ExpiresAt = expiresAt,
            Status = "active",
            LastNfcCounter = 0,
        };
        // Give the brand-new card an initial mobile PIN so the holder can sign in
        // immediately (without this, PinHash stays null and sign-in-with-pin
        // always fails for a just-issued card). Returned once to the issuing
        // officer to relay to the citizen; only the bcrypt hash persists.
        var pin = AssignNewPin(card);
        db.DigitalIdCards.Add(card);

        try { await db.SaveChangesAsync(ct); }
        catch (DbUpdateException ex) when (IsUnique(ex))
        {
            throw SarhException.Conflict(
                "تعارض في رقم البطاقة أو الرقم الرقمي.",
                "Conflict on card_serial or digital_id_number.");
        }

        var keys = await keyStore.MintForCardAsync(card.Id, ct);

        db.IdIssuanceHistory.Add(new IdIssuanceHistory
        {
            Id = Guid.NewGuid(),
            CitizenId = dto.CitizenId,
            CardId = card.Id,
            Action = "issued",
            Reason = null,
            OfficerId = actor.OfficerId,
        });
        await db.SaveChangesAsync(ct);

        // Issue the DigitalId VC into the citizen's SSI wallet and stamp the
        // wallet DID on the card. Best-effort — an SSI outage falls back to a
        // placeholder DID and never fails issuance.
        await IssueDigitalIdVcAsync(card, ct);
        await db.SaveChangesAsync(ct);

        await notifications.NotifyCitizenAsync(
            card.CitizenId,
            "تم إصدار بطاقة الهوية الرقمية",
            $"تم إصدار بطاقتك برقم {card.DigitalIdNumber}.",
            new { card_id = card.Id, digital_id_number = card.DigitalIdNumber },
            ct, alsoSms: true);

        return new IssueCardResult
        {
            Card = CardView.From(card),
            NfcKeys = new IssueCardNfcKeys
            {
                MetaReadKeyHex = Convert.ToHexString(keys.MetaReadKey),
                SdmFileReadKeyHex = Convert.ToHexString(keys.SdmFileReadKey),
                KmsKeyId = NfcKeyStoreService.LocalKmsKeyId,
            },
            SunUrlTemplate = SunUrlTemplate(),
            Pin = pin,
        };
    }

    // ---------------- FREEZE / REVOKE ----------------
    public Task<CardView> FreezeAsync(Guid cardId, FreezeCardDto dto, CurrentUser actor, CancellationToken ct)
        => TransitionAsync(cardId, "frozen", dto.Reason, actor, "frozen", ct);

    public Task<CardView> RevokeAsync(Guid cardId, RevokeCardDto dto, CurrentUser actor, CancellationToken ct)
        => TransitionAsync(cardId, "revoked", dto.Reason, actor, "revoked", ct);

    // ---------------- REISSUE ----------------
    public async Task<IssueCardResult> ReissueAsync(Guid cardId, ReissueCardDto dto, CurrentUser actor, CancellationToken ct)
    {
        if (actor.OfficerId is null) throw SarhException.Forbidden();

        var old = await db.DigitalIdCards.AsNoTracking().FirstOrDefaultAsync(c => c.Id == cardId, ct)
            ?? throw SarhException.NotFound("البطاقة", "Card");

        var citizen = await db.Citizens.AsNoTracking().FirstOrDefaultAsync(c => c.Id == old.CitizenId, ct)
            ?? throw SarhException.NotFound("المواطن", "Citizen");

        // notify:false — reissue sends its own "card reissued" message below; we
        // don't want the holder to also receive a scary "card revoked" message.
        await TransitionAsync(cardId, "revoked", $"إعادة إصدار: {dto.Reason}", actor, "revoked", ct, notify: false);

        var year = DateTime.UtcNow.Year;
        var region = ParseRegionFromDigitalId(old.DigitalIdNumber);
        // Default-mint-new: digital_id_number is NOT NULL UNIQUE at the DB
        // level, and the old card row stays around (revoked) for audit, so
        // reusing the same number would 409. Only keep the old number when
        // the caller is explicit about it AND has freed it first (advanced
        // use; not exercised by default).
        var digitalIdNumber = dto.KeepDigitalIdNumber == true
            ? old.DigitalIdNumber
            : await numbers.NextAsync(region, year, ct);

        var cardSerial = $"LY-{RandomHexUpper(12)}";
        var card = new DigitalIdCard
        {
            Id = Guid.NewGuid(),
            CitizenId = old.CitizenId,
            DigitalIdNumber = digitalIdNumber,
            CardSerial = cardSerial,
            PhotoHash = old.PhotoHash,
            DataHash = IdentityHash.Compute(citizen, digitalIdNumber),
            IssuedByOfficerId = actor.OfficerId,
            IssuedAt = DateTimeOffset.UtcNow,
            ExpiresAt = DateTimeOffset.UtcNow.AddYears(5),
            Status = "active",
            LastNfcCounter = 0,
        };
        // A reissued card is a fresh card and needs its own PIN, otherwise the
        // holder can't sign in after a reissue.
        var pin = AssignNewPin(card);
        db.DigitalIdCards.Add(card);

        try { await db.SaveChangesAsync(ct); }
        catch (DbUpdateException ex) when (IsUnique(ex))
        {
            throw SarhException.Conflict(
                "تعارض في رقم البطاقة الجديد.",
                "Conflict on reissued card_serial / digital_id_number.");
        }

        var keys = await keyStore.MintForCardAsync(card.Id, ct);

        db.IdIssuanceHistory.Add(new IdIssuanceHistory
        {
            Id = Guid.NewGuid(),
            CitizenId = old.CitizenId,
            CardId = card.Id,
            Action = "re-issued",
            Reason = dto.Reason,
            OfficerId = actor.OfficerId,
        });
        await db.SaveChangesAsync(ct);

        await IssueDigitalIdVcAsync(card, ct);
        await db.SaveChangesAsync(ct);

        await notifications.NotifyCitizenAsync(
            card.CitizenId,
            "تم إعادة إصدار بطاقتك الرقمية",
            $"تم إصدار بطاقة هوية رقمية جديدة لك برقم {card.DigitalIdNumber}. أصبحت بطاقتك السابقة لاغية.",
            new { card_id = card.Id, digital_id_number = card.DigitalIdNumber, reissued = true },
            ct, alsoSms: true);

        return new IssueCardResult
        {
            Card = CardView.From(card),
            NfcKeys = new IssueCardNfcKeys
            {
                MetaReadKeyHex = Convert.ToHexString(keys.MetaReadKey),
                SdmFileReadKeyHex = Convert.ToHexString(keys.SdmFileReadKey),
                KmsKeyId = NfcKeyStoreService.LocalKmsKeyId,
            },
            SunUrlTemplate = SunUrlTemplate(),
            Pin = pin,
        };
    }

    // ---------------- helpers ----------------
    private async Task<CardView> TransitionAsync(
        Guid cardId, string nextStatus, string reason, CurrentUser actor,
        string historyAction, CancellationToken ct, bool notify = true)
    {
        if (actor.OfficerId is null) throw SarhException.Forbidden();

        var card = await db.DigitalIdCards.FirstOrDefaultAsync(c => c.Id == cardId, ct)
            ?? throw SarhException.NotFound("البطاقة", "Card");

        if (card.Status == "revoked")
            throw SarhException.Conflict(
                "البطاقة مُلغاة بالفعل ولا يمكن تعديل حالتها.",
                "Card is already revoked.");
        if (nextStatus == "frozen" && card.Status == "frozen")
            throw SarhException.Conflict("البطاقة مجمّدة بالفعل.", "Card is already frozen.");

        card.Status = nextStatus;
        if (nextStatus == "revoked")
        {
            card.RevokedAt = DateTimeOffset.UtcNow;
            card.RevokedReason = reason;
        }

        db.IdIssuanceHistory.Add(new IdIssuanceHistory
        {
            Id = Guid.NewGuid(),
            CitizenId = card.CitizenId,
            CardId = card.Id,
            Action = historyAction,
            Reason = reason,
            OfficerId = actor.OfficerId,
        });

        await db.SaveChangesAsync(ct);

        if (!notify) return CardView.From(card);

        // The card holder must be told about a status change to their identity
        // card. Revocation is a critical security event → also push an SMS.
        var (titleAr, bodyAr) = nextStatus == "frozen"
            ? ("تم تجميد بطاقة الهوية الرقمية",
               $"تم تجميد بطاقتك رقم {card.DigitalIdNumber} مؤقتاً." + (string.IsNullOrWhiteSpace(reason) ? "" : $" السبب: {reason}"))
            : ("تم إلغاء بطاقة الهوية الرقمية",
               $"تم إلغاء بطاقتك رقم {card.DigitalIdNumber}." + (string.IsNullOrWhiteSpace(reason) ? "" : $" السبب: {reason}"));
        await notifications.NotifyCitizenAsync(
            card.CitizenId, titleAr, bodyAr,
            new { card_id = card.Id, status = nextStatus, reason },
            ct, alsoSms: nextStatus == "revoked");

        return CardView.From(card);
    }

    private string ResolvePhotoHash(IssueCardDto dto, string? citizenPhotoPath)
    {
        if (!string.IsNullOrEmpty(dto.PhotoSha256))
        {
            if (!PhotoSha256Re.IsMatch(dto.PhotoSha256))
                throw SarhException.Validation(
                    "بصمة الصورة غير صالحة (يجب أن تكون 64 حرفاً سادس عشر).",
                    "photo_sha256 must be 64 hex characters.");
            return dto.PhotoSha256.ToLowerInvariant();
        }

        var path = dto.PhotoPath ?? citizenPhotoPath;
        if (string.IsNullOrEmpty(path))
            throw SarhException.Validation(
                "يجب توفير صورة المواطن أو بصمتها قبل إصدار البطاقة.",
                "Either photo_path or photo_sha256 is required.");

        // Local-FS storage hashing lands in Phase 5; until then we deterministically
        // hash the storage path so the column has a stable value.
        log.LogInformation("Photo hash placeholder for path {Path} until Phase 5 storage land", path);
        var hash = SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(path));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    // Issues the DigitalId verifiable credential into the citizen's SSI wallet
    // and stamps the wallet DID onto the card. Best-effort: any SSI failure
    // degrades to a placeholder DID so card issuance always completes.
    private async Task IssueDigitalIdVcAsync(DigitalIdCard card, CancellationToken ct)
    {
        try
        {
            var vc = await ssi.IssueDigitalIdVcAsync(card.Id, ct);
            if (vc is not null) card.Did = vc.Did;
            else AttachPlaceholderVc(card);
        }
        catch (Exception ex)
        {
            log.LogWarning(ex, "SSI DigitalId VC issuance failed for card {CardId}; using placeholder DID.", card.Id);
            AttachPlaceholderVc(card);
        }

        // The SSI DID is the citizen's stable wallet DID, so every card the
        // citizen has ever held resolves to the SAME value — but
        // digital_id_cards.did is UNIQUE (ux_did_cards_did). Keep the DID on the
        // citizen's current card only: release it from any prior (e.g. revoked)
        // card before the caller persists it here, otherwise a reissue collides
        // with the old card's DID and the save 500s. Placeholder DIDs are
        // per-card unique, so this is a harmless no-op for them.
        if (!string.IsNullOrEmpty(card.Did))
            await db.Database.ExecuteSqlRawAsync(
                "UPDATE digital_id_cards SET did = NULL WHERE citizen_id = {0} AND id <> {1} AND did = {2}",
                new object[] { card.CitizenId, card.Id, card.Did }, ct);
    }

    private void AttachPlaceholderVc(DigitalIdCard card)
    {
        card.Did = $"did:placeholder:LY:{Guid.NewGuid():N}";
    }

    private string SunUrlTemplate()
    {
        var baseUrl = config["Sarh:NfcSunBaseUrl"]
            ?? Environment.GetEnvironmentVariable("NFC_SUN_BASE_URL")
            ?? "https://verify.sarh.ly/v";
        return $"{baseUrl}?p={{picc}}&c={{cmac}}";
    }

    private static string RandomHexUpper(int bytes) =>
        Convert.ToHexString(RandomNumberGenerator.GetBytes(bytes));

    private static string ParseRegionFromDigitalId(string id)
    {
        var m = Regex.Match(id, "^LY-([0-9]{2,4})-");
        if (!m.Success) throw SarhException.Upstream($"Cannot parse region from digital ID: {id}");
        return m.Groups[1].Value;
    }

    private const int UNIQUE_VIOLATION = 2627;
    private const int UNIQUE_VIOLATION_INDEX = 2601;

    private static bool IsUnique(DbUpdateException ex) =>
        ex.InnerException is Microsoft.Data.SqlClient.SqlException se &&
        (se.Number == UNIQUE_VIOLATION || se.Number == UNIQUE_VIOLATION_INDEX);
}
